File: app/src/main/java/uk/co/ourfriendirony/medianotifier/mediaitem/artist/Release.java
package uk.co.ourfriendirony.medianotifier.mediaitem.artist;

import static uk.co.ourfriendirony.medianotifier.general.Helper.getColumnValue;
import static uk.co.ourfriendirony.medianotifier.general.Helper.stringToDate;

import android.database.Cursor;
import android.util.Log;

import androidx.annotation.NonNull;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import uk.co.ourfriendirony.medianotifier.clients.musicbrainz.artist.get.ArtistGetReleaseGroup;
import uk.co.ourfriendirony.medianotifier.db.artist.ArtistDatabaseDefinition;
import uk.co.ourfriendirony.medianotifier.mediaitem.MediaItem;

public class Release implements MediaItem {
    private final String id;
    private final String title;
    private final Date releaseDate;
    // TODO: fully implement played as an item
    private final boolean played = false;
    private String subid = "";
    private String subtitle = "";
    private String description = "";
    private String externalUrl;
    private List<MediaItem> children = new ArrayList<>();

    public Release(ArtistGetReleaseGroup release, Artist artist) {
        this.id = artist.getId();
        this.subid = release.getId();
        if (!release.getDisambiguation().isEmpty()) {
            this.title = release.getTitle() + " (" + release.getDisambiguation() + ")";
        } else {
            this.title = release.getTitle();
        }
        this.subtitle = artist.getTitle();
        this.releaseDate = release.getFirstReleaseDate();
        Log.d("[API GET]", this.toString());
    }

    public Release(Cursor cursor) {
        this.id = getColumnValue(cursor, ArtistDatabaseDefinition.ID);
        this.subid = getColumnValue(cursor, ArtistDatabaseDefinition.SUBID);
        this.title = getColumnValue(cursor, ArtistDatabaseDefinition.TITLE);
        this.subtitle = getColumnValue(cursor, ArtistDatabaseDefinition.SUBTITLE);
        this.description = getColumnValue(cursor, ArtistDatabaseDefinition.DESCRIPTION);
        this.releaseDate = stringToDate(getColumnValue(cursor, ArtistDatabaseDefinition.RELEASE_DATE));
        this.externalUrl = getColumnValue(cursor, ArtistDatabaseDefinition.EXTERNAL_URL);
        Log.d("[DB READ]", this.toString());
    }

    @Override
    public String getId() {
        return id;
    }

    @Override
    public String getSubId() {
        return subid;
    }

    @Override
    public String getTitle() {
        return title;
    }

    @Override
    public String getSubtitle() {
        return subtitle;
    }

    @Override
    public String getDescription() {
        return description;
    }

    @Override
    public Date getReleaseDate() {
        return releaseDate;
    }

    @Override
    public String getReleaseDateFull() {
        if (releaseDate != null) {
            return new SimpleDateFormat("dd/MM/yyyy", Locale.UK).format(releaseDate);
        }
        return MediaItem.NO_DATE;
    }

    @Override
    public String getReleaseDateYear() {
        if (releaseDate != null) {
            return new SimpleDateFormat("yyyy", Locale.UK).format(releaseDate);
        }
        return MediaItem.NO_DATE;
    }

    @Override
    public List<MediaItem> getChildren() {
        return children;
    }

    @Override
    public void setChildren(List<MediaItem> children) {
        this.children = children;
    }

    @Override
    public int countChildren() {
        return children.size();
    }

    @Override
    public String getExternalLink() {
        return externalUrl;
    }

    @Override
    public Boolean getPlayed() {
        return played;
    }

    @NonNull
    public String toString() {
        return "Release: " + getTitle() + " > " + getReleaseDateFull();
    }
}

File: app/src/main/kotlin/uk/co/ourfriendirony/medianotifier/mediaitem/artist/Release.kt
package uk.co.ourfriendirony.medianotifier.mediaitem.artist

import android.database.Cursor
import android.util.Log
import uk.co.ourfriendirony.medianotifier.clients.musicbrainz.release.get.ReleaseGet
import uk.co.ourfriendirony.medianotifier.clients.musicbrainz.release.get.ReleaseGetTrack
import uk.co.ourfriendirony.medianotifier.clients.musicbrainz.releasegroup.get.ReleaseGroupGet
import uk.co.ourfriendirony.medianotifier.db.artist.ArtistDatabaseDefinition
import uk.co.ourfriendirony.medianotifier.general.Helper.getColumnValue
import uk.co.ourfriendirony.medianotifier.general.Helper.stringToDate
import uk.co.ourfriendirony.medianotifier.mediaitem.MediaItem
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class Release : MediaItem {
    override val id: String
    override val title: String?
    override val releaseDate: Date?
    override val isParent = false

    // TODO: fully implement played as an item
    override val played = false
    override var subId: String? = ""
        private set
    override var subtitle: String? = ""
        private set
    override var description: String? = ""
        private set
    override var externalLink: String? = null
        private set
    override var children: MutableList<MediaItem> = ArrayList()

    constructor(release: ReleaseGet, releaseGroup: ReleaseGroupGet, artist: Artist) {
        id = artist.id
        subId = releaseGroup.id
        title = if (releaseGroup.disambiguation!!.isNotEmpty()) {
            releaseGroup.title + " (" + releaseGroup.disambiguation + ")"
        } else {
            releaseGroup.title
        }
        subtitle = artist.title
        releaseDate = release.date
        description = getTracksAsDescription(release.media?.get(0)?.tracks)

        Log.d("[API GET]", this.toString())
    }

    constructor(cursor: Cursor?) {
        id = getColumnValue(cursor!!, ArtistDatabaseDefinition.ID)
        subId = getColumnValue(cursor, ArtistDatabaseDefinition.SUBID)
        title = getColumnValue(cursor, ArtistDatabaseDefinition.TITLE)
        subtitle = getColumnValue(cursor, ArtistDatabaseDefinition.SUBTITLE)
        description = getColumnValue(cursor, ArtistDatabaseDefinition.DESCRIPTION)
        releaseDate = stringToDate(getColumnValue(cursor, ArtistDatabaseDefinition.RELEASE_DATE))
        externalLink = getColumnValue(cursor, ArtistDatabaseDefinition.EXTERNAL_URL)
        Log.d("[DB READ]", this.toString())
    }

    override val releaseDateFull: String
        get() = if (releaseDate != null) {
            SimpleDateFormat("dd/MM/yyyy", Locale.UK).format(releaseDate)
        } else MediaItem.NO_DATE
    override val releaseDateYear: String
        get() = if (releaseDate != null) {
            SimpleDateFormat("yyyy", Locale.UK).format(releaseDate)
        } else MediaItem.NO_DATE

    private fun getTracksAsDescription(tracks: List<ReleaseGetTrack>?): String {
        var d = ""
        tracks?.forEach { track ->
            d += "${track.position.toString()}:${track.title}\n"

        }
        return d
    }

    override fun countChildren(): Int {
        return children.size
    }

    override fun toString(): String {
        return "Release: $title > $releaseDateFull"
    }
}